akka-stream-tests/src/test/scala/akka/stream/impl/fusing/InterpreterSupervisionSpec.scala

```diff
@@ -0,0 +1,656 @@
+      lastEvents() should be(Set(RequestOne, OnNext(1)))
+
+      downstream.requestOne()
+      lastEvents() should be(Set(OnNext(-1)))
+
+      downstream.requestOne()
+      lastEvents() should be(Set(OnNext(-1)))
+
+      upstream.onNext(2) // boom
+      lastEvents() should be(Set(RequestOne))
+
+      downstream.requestOne()
+      lastEvents() should be(Set(OnNext(-1)))
+
+      upstream.onNext(3)
+      lastEvents() should be(Set.empty)
+
+      downstream.requestOne()
+      lastEvents() should be(Set(RequestOne, OnNext(3)))
+    }
+
+    "restart when Expand `extrapolate` throws" in new TestSetup(Seq(Expand(
+      (in: Int) ⇒ in,
+      (agg: Int) ⇒ if (agg == 2) throw TE else (agg, -math.abs(agg)),
+      restartingDecider))) {
+
+      lastEvents() should be(Set(RequestOne))
+
+      upstream.onNext(1)
+      lastEvents() should be(Set.empty)
+
+      downstream.requestOne()
+      lastEvents() should be(Set(RequestOne, OnNext(1)))
+
+      downstream.requestOne()
+      lastEvents() should be(Set(OnNext(-1)))
+
+      upstream.onNext(2) // boom
+      lastEvents() should be(Set.empty)
+
+      downstream.requestOne()
+      lastEvents() should be(Set(RequestOne))
+
+      upstream.onNext(3)
+      lastEvents() should be(Set.empty)
+
+      downstream.requestOne()
+      lastEvents() should be(Set(RequestOne, OnNext(3)))
+
+      downstream.requestOne()
+      lastEvents() should be(Set(OnNext(-3)))
+    }
+
+    "resume when Expand `extrapolate` throws" in new TestSetup(Seq(Expand(
+      (in: Int) ⇒ in,
+      (agg: Int) ⇒ if (agg == 2) throw TE else (agg, -math.abs(agg)),
+      resumingDecider))) {
+
+      pending // see blow FIXME
+
+      lastEvents() should be(Set(RequestOne))
+
+      upstream.onNext(1)
+      lastEvents() should be(Set.empty)
+
+      downstream.requestOne()
+      lastEvents() should be(Set(RequestOne, OnNext(1)))
+
+      downstream.requestOne()
+      lastEvents() should be(Set(OnNext(-1)))
+
+      upstream.onNext(2) // boom
+      lastEvents() should be(Set.empty)
+
+      downstream.requestOne() // FIXME this throws IllegalStateException: Cannot pull while holding, only pushAndPull
+      lastEvents() should be(Set(OnNext(-1)))
+
+      upstream.onNext(3)
+      lastEvents() should be(Set.empty)
+
+      downstream.requestOne()
+      lastEvents() should be(Set(RequestOne, OnNext(3)))
+
+      downstream.requestOne()
+      lastEvents() should be(Set(OnNext(-3)))
+    }
+
+    "resume when onPull throws before pushing all buffered elements" in new TestSetup(Seq(
+      BufferingTestStage(resumingDecider))) {
+
+      downstream.requestOne()
+      lastEvents() should be(Set(RequestOne))
+      upstream.onNext(1)
+      lastEvents() should be(Set(OnNext(1)))
+
+      downstream.requestOne()
+      lastEvents() should be(Set(OnNext(2)))
+
+      downstream.requestOne()
+      // 3 => boom
+      // FIXME feels wrong? it should deliver OnNext(4)?
+      lastEvents() should be(Set(RequestOne))
```

Comment:
no, if the resume strategy pulls. This is what I said regarding one-to-many stages. The stage expects a pull from downstream to serve the next element from the list, but instead the resume strategy pulls, so it gets a push from upstream! This can lead to serious violation of expectations in the stage.